### PR TITLE
ApplyStartupHook diagnostic IPC command

### PR DESCRIPTION
**Pull request by jander-msft**
These changes allow adding diagnostic startup hooks through a new diagnostic IPC command. The new command has a single string property on it that represents the path to the assembly on disk that contains the startup hook.

The diagnostic startup hooks are stored in a static variable and then passed to `StartupHookProvider.ManagedStartup`; here the list of hooks is combined with the ones from the `STARTUP_HOOKS` app context data and then processed as startup hooks normally would be processed. Tests for both the managed library changes and the event pipe have been added.

I did not implement this in Mono; I did not see a way of dynamically providing startup hooks (e.g. such as environment variable) and didn't want to enable this dynamic notion without a scenario in place. I also have very little understanding on how to implement it for that runtime, so I've fixed it up enough that it would pass an empty string for the diagnostic startup hooks, thus preserving the existing behavior.

I did not implement this for AOT scenarios since adding startup hooks there doesn't make sense (no managed code execution).

Corresponding diagnostics change: https://github.com/dotnet/diagnostics/pull/3918

Partially addresses https://github.com/dotnet/runtime/issues/83756

cc @dotnet/dotnet-monitor

### CODE REVIEW

**Comment by jander-msft** [comment]
cc @vitek-karas and @elinor-fung: wanted your feedback on where this new property should be get/set from the diagnostic command (currently, it's going through the host runtime contract), and how to compose that (and maybe the other host properties) with the list of properties that were passed into `coreclr_initialize` (which is looks to be the same set as those from the host runtime contract, but may be different depending on how the caller invokes `coreclr_intialize`). Any other thoughts around structures to use, tests that could be added, etc... would be appreciated.

File: src/coreclr/dlls/mscoree/exports.cpp
Comment (jander-msft) [inline src/coreclr/dlls/mscoree/exports.cpp:209]: What I'm trying to do in this function is create a new property list of the app domain creation that composes the original property list passed into `coreclr_intialize` with the new `DIAGNOSTIC_STARTUP_HOOKS` property from the host. My thought is that if I load this property from the property list (if it exists) and set it into the host, allow the diagnostic command to modify it, and then retrieve it again for recomposition, then it'll be the most update to date without having lost any information. I'm not trying to enumerate all properties from the host because that's not what the behavior was before (well, it's the observed behavior because the host is passing the same properties into `coreclr_intialize` but that's not know from this function).

Comment (jkotas) [reply src/coreclr/dlls/mscoree/exports.cpp:209]: Does this need to go through the AppContext property bag at all? It looks unnecessarily complex.

I think it would be a lot simpler for the eventpipe diagnostic command to store the path in a static variable, and the startup hook provider to read it from that static variable.

Comment (jander-msft) [reply src/coreclr/dlls/mscoree/exports.cpp:209]: How would I get the startup hook provider to read the static variable? Is the static variable native or managed (I assume it's native and then supplied to `StartupHookProvider` in some way)? Is changing the signature of `StartupHookProvider.MamagedStartup` permitted to take a string parameter? If I'm modifying any managed signatures in `StartupHookProvider`, do I need to make sure Mono works with it?

Comment (jkotas) [reply src/coreclr/dlls/mscoree/exports.cpp:209]: > Is the static variable native or managed (I assume it's native and then supplied to StartupHookProvider in some way)?

Right, it would be native and supplied to `StartupHookProvider`.

> Is changing the signature of StartupHookProvider.ManagedStartup permitted to take a string parameter?

Yes, that's fine. The easiest way to do it is to change it to `void ManagedStartup(char* pDiagnosticStartupHook)` and then create the managed string out of `pDiagnosticStartupHook`.

> I need to make sure Mono works with it?

Yes. (It should not be hard - you can leave it as the last thing to fix.)

**Comment by jander-msft** [comment]
ILLink supression needs to be updated to remove the trim warnings: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Suppressions.LibraryBuild.xml#L22

File: src/coreclr/System.Private.CoreLib/src/System/StartupHookProvider.CoreCLR.cs
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/StartupHookProvider.CoreCLR.cs:23]:
```suggestion
                ProcessStartupHooks(new string(pDiagnosticStartupHooks));
```
`new string` returns `string.Empty` for `(char*)null`

File: src/libraries/System.Private.CoreLib/src/System/StartupHookProvider.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/StartupHookProvider.cs:36]: This should avoid allocating the `List` if there are no startup hooks (the most common case).

**Comment by jkotas** [comment]
Test failures:
```
  Starting:    Loader.StartupHooks.XUnitWrapper (parallel test collections = on, max threads = 6)
    Loader/StartupHooks/StartupHookTests/StartupHookTests.sh [FAIL]
      Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
         at System.String.MakeSeparatorListVectorized(System.ReadOnlySpan`1<Char>, System.Collections.Generic.ValueListBuilder`1<Int32> ByRef, Char, Char, Char)
         at System.String.MakeSeparatorListAny(System.ReadOnlySpan`1<Char>, System.ReadOnlySpan`1<Char>, System.Collections.Generic.ValueListBuilder`1<Int32> ByRef)
         at System.String.SplitInternal(System.ReadOnlySpan`1<Char>, Int32, System.StringSplitOptions)
         at System.StartupHookProvider.ProcessStartupHooks(System.String)
         at StartupHookTests.ValidHookName()
```

File: src/coreclr/vm/assembly.cpp
Comment (vitek-karas) [inline src/coreclr/vm/assembly.cpp:55]: Why not make this into a static variable local to this file. The only new member on the class would be the AddDiagnosticStartupHook. I would try to avoid adding private members to classes unless necessary, it has no real benefit and it makes compilation/maintenance harder as any changes have to touch two files (and touching the header trigger rebuilding of lot of dependencies).

**Review by vitek-karas — COMMENTED** [review]
We should really try to add some test for this new capability. Hopefully we already have some test infra somewhere to help with the setup of the diag channel.

**Comment by jander-msft** [comment]
> We should really try to add some test for this new capability. Hopefully we already have some test infra somewhere to help with the setup of the diag channel.

There seems to be event pipe tests located [here](https://github.com/dotnet/runtime/tree/main/src/tests/tracing/eventpipe); I don't think any of them involve loading an assembly separate from the entrypoint assembly, so there may be additional work for enabling that compilation and locating of that assembly in the test output.

I can also add some variants of the existing tests [here](https://github.com/dotnet/runtime/blob/main/src/tests/Loader/StartupHooks/StartupHookTests.cs) for the managed side.

**Comment by jander-msft** [comment]
> > We should really try to add some test for this new capability. Hopefully we already have some test infra somewhere to help with the setup of the diag channel.
> 
> There seems to be event pipe tests located [here](https://github.com/dotnet/runtime/tree/main/src/tests/tracing/eventpipe); I don't think any of them involve loading an assembly separate from the entrypoint assembly, so there may be additional work for enabling that compilation and locating of that assembly in the test output.

If I recall correctly, some of the testing for these diagnostic commands is covered in the dotnet/diagnostics repo. I'll ask @davmason to chime in where (either in runtime or diagnostics) he thinks adding tests for the diagnostic command is appropriate.

**Comment by jkotas** [comment]
The test is timing out:
```
  Starting:    tracing.eventpipe.XUnitWrapper (parallel test collections = on, max threads = 2)
   tracing.eventpipe.XUnitWrapper: [Long Running Test] 'tracing/eventpipe/applystartuphook/applystartuphook/applystartuphook.sh', Elapsed: 00:02:34
   tracing.eventpipe.XUnitWrapper: [Long Running Test] 'tracing/eventpipe/applystartuphook/applystartuphook/applystartuphook.sh', Elapsed: 00:04:34
   tracing.eventpipe.XUnitWrapper: [Long Running Test] 'tracing/eventpipe/applystartuphook/applystartuphook/applystartuphook.sh', Elapsed: 00:06:34
   tracing.eventpipe.XUnitWrapper: [Long Running Test] 'tracing/eventpipe/applystartuphook/applystartuphook/applystartuphook.sh', Elapsed: 00:08:34
Attempting to collect crash dump: /home/helixbot/dotnetbuild/dumps/crashdump_2037.dmp
```

**Comment by jkotas** [comment]
The change looks good to me overall. @davmason should review the test and the event pipe plumbing.

**Comment by davmason** [comment]
The test is failing because it doesn't find the hook dll

```
0:000> !pe -nested
Exception object: 000001a69f40bfb0
Exception type:   System.ArgumentException
Message:          Startup hook assembly 'C:\h\w\9BE60914\w\9C370888\e\tracing\eventpipe\applystartuphook\applystartuphook\hooks\Basic.dll' failed to load. See inner exception for details.
InnerException:   System.IO.FileNotFoundException, Use !PrintException 000001A69F408F98 to see more.
StackTrace (generated):
    SP               IP               Function
    000000160FB7B2C0 00007FF8C0C48C52 System_Private_CoreLib!System.StartupHookProvider.CallStartupHook(StartupHookNameOrPath)+0x342
    000000160FB7E6E0 00007FF8C0C48810 System_Private_CoreLib!System.StartupHookProvider.ProcessStartupHooks(System.String)+0x300

StackTraceString: <none>
HResult: 80070057

Nested exception -------------------------------------------------------------
Exception object: 000001a69f408f98
Exception type:   System.IO.FileNotFoundException
Message:          Could not load file or assembly 'C:\h\w\9BE60914\w\9C370888\e\tracing\eventpipe\applystartuphook\applystartuphook\hooks\Basic.dll'. The system cannot find the path specified.
InnerException:   <none>
StackTrace (generated):
    SP               IP               Function
    000000160FB7E4F0 00007FF8C0D8976E System_Private_CoreLib!System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr, System.String, System.String, System.Runtime.CompilerServices.ObjectHandleOnStack)+0x7e
    000000160FB7E5E0 00007FF8C0D8AD70 System_Private_CoreLib!System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(System.String)+0xb0
    000000160FB7E650 00007FF8C0C48978 System_Private_CoreLib!System.StartupHookProvider.CallStartupHook(StartupHookNameOrPath)+0x68

StackTraceString: <none>
HResult: 80070003
```

**Review by davmason — APPROVED** [review]
Once the tests are passing it looks good to me. Thanks Justin!

File: src/tests/tracing/eventpipe/applystartuphook/ApplyStartupHookValidation.cs
Comment (jander-msft) [inline src/tests/tracing/eventpipe/applystartuphook/ApplyStartupHookValidation.cs:49]: This is throwing an exception when running on Mono because it's not supported:
```
Microsoft.Diagnostics.NETCore.Client.ServerErrorException: ApplyStartupHook - Not supported by this runtime.
         at Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.ValidateResponseMessage(IpcMessage responseMessage, String operationName, ValidateResponseOptions options)
         at Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.ApplyStartupHook(String startupHookPath)
         at Tracing.Tests.ApplyStartupHookValidation.ApplyStartupHookValidation.<>c__DisplayClass0_0.<<TEST_ApplyStartupHookAtStartupSuspension>b__0>d.MoveNext()
```

Is there a way to detect whether this is running in Mono vs CoreClr (then I can test that it does throw the expected exception on Mono) or possibly to only run this test for CoreClr?

Comment (elinor-fung) [reply src/tests/tracing/eventpipe/applystartuphook/ApplyStartupHookValidation.cs:49]: For detection at run-time, you can add a reference to `CoreCLRTestLibrary` and use `TestLibrary.Utilities.IsMonoRuntime`:
https://github.com/dotnet/runtime/blob/fc486b4087bd3a6c6c2934da8b3c22cb4bf73523/src/tests/Common/CoreCLRTestLibrary/Utilities.cs#L95

To disable a test entirely, you can update the project to be disabled based on the runtime. For example:
https://github.com/dotnet/runtime/blob/fc486b4087bd3a6c6c2934da8b3c22cb4bf73523/src/tests/tracing/runtimeeventsource/runtimeeventsource.csproj#L10

Comment (elinor-fung) [reply src/tests/tracing/eventpipe/applystartuphook/ApplyStartupHookValidation.cs:49]: I'd also mark the test nativeaot-incompatible:
```xml
<!-- Startup hooks are not supported -->
<NativeAotIncompatible>true</NativeAotIncompatible>
```

**Comment by jander-msft** [comment]
Seems the test failures are all non-Windows runs. I'm going to try to add more logging in the test to narrow the problem space and try to get a Linux machine and debug it.

**Comment by elinor-fung** [comment]
Not sure if it is relevant, but this seems to be the only test that alternates between `ReverseServer` and `DiagnosticsClient` to communicate with the target process - maybe something in the test framework doesn't handle using them together?

**Comment by jander-msft** [comment]
> Not sure if it is relevant, but this seems to be the only test that alternates between `ReverseServer` and `DiagnosticsClient` to communicate with the target process - maybe something in the test framework doesn't handle using them together?

Yeah, I'm going to try to use the reverse server connection for the startup hook application. If that works, that means there's likely a pre-existing bug in the diagnostics handlers within the runtime that can't handle communicating on both the reverse and forward connections.

File: src/tests/tracing/eventpipe/applystartuphook/ApplyStartupHookValidation.cs
Comment (jander-msft) [reply src/tests/tracing/eventpipe/applystartuphook/ApplyStartupHookValidation.cs]: I ended up excluding running this test on Mono.

**Comment by jander-msft** [comment]
> > Not sure if it is relevant, but this seems to be the only test that alternates between `ReverseServer` and `DiagnosticsClient` to communicate with the target process - maybe something in the test framework doesn't handle using them together?
> 
> Yeah, I'm going to try to use the reverse server connection for the startup hook application. If that works, that means there's likely a pre-existing bug in the diagnostics handlers within the runtime that can't handle communicating on both the reverse and forward connections.

Using only the reverse connection seems to work on some of the Helix machines regardless of OS, at least in the test legs that actually completed. I'll rerun the other legs once the test queues are back to normal.

**Comment by elinor-fung** [comment]
> > > Not sure if it is relevant, but this seems to be the only test that alternates between `ReverseServer` and `DiagnosticsClient` to communicate with the target process - maybe something in the test framework doesn't handle using them together?
> > 
> > 
> > Yeah, I'm going to try to use the reverse server connection for the startup hook application. If that works, that means there's likely a pre-existing bug in the diagnostics handlers within the runtime that can't handle communicating on both the reverse and forward connections.
> 
> Using only the reverse connection seems to work on some of the Helix machines regardless of OS, at least in the test legs that actually completed. I'll rerun the other legs once the test queues are back to normal.

@davmason does this sound familiar / known to you?

**Comment by jander-msft** [comment]
> > > > Not sure if it is relevant, but this seems to be the only test that alternates between `ReverseServer` and `DiagnosticsClient` to communicate with the target process - maybe something in the test framework doesn't handle using them together?
> > > 
> > > 
> > > Yeah, I'm going to try to use the reverse server connection for the startup hook application. If that works, that means there's likely a pre-existing bug in the diagnostics handlers within the runtime that can't handle communicating on both the reverse and forward connections.
> > 
> > 
> > Using only the reverse connection seems to work on some of the Helix machines regardless of OS, at least in the test legs that actually completed. I'll rerun the other legs once the test queues are back to normal.
> 
> @davmason does this sound familiar / known to you?

Sorry, I was trying to indicate that for the test legs that completed, they passed. The others didn't complete due to infra issues with the Helix queues.

But I do think it's strange that in the [original test implementation](https://github.com/dotnet/runtime/pull/86813/commits/d970ae6f701ccce430b0615e40b4fda2d0a9655a), that sending the ApplyStartupHook command over the default channel and handling the accept and resume on the reverse channel caused it to hang when trying to accept the second incoming reverse connection.

It appears that performing all commands over the same channel works; this is typically how the dotnet-* diagnostic tools interact with the target process.

**Comment by elinor-fung** [comment]
> But I do think it's strange that in the [original test implementation](https://github.com/dotnet/runtime/pull/86813/commits/d970ae6f701ccce430b0615e40b4fda2d0a9655a), that sending the ApplyStartupHook command over the default channel and handling the accept and resume on the reverse channel caused it to hang when trying to accept the second incoming reverse connection.

I'd agree (but am not too familiar with the internal workings there). It was also only non-Windows (so using sockets vs named pipes), right? Could you file an issue? Unless @davmason thinks this is expected.

**Comment by davmason** [comment]
> I'd agree (but am not too familiar with the internal workings there). It was also only non-Windows (so using sockets vs named pipes), right? Could you file an issue? Unless @davmason thinks this is expected.

I need a little bit to investigate, I don't know off the top of my head. I will let you know later today if it's expected or not.

**Comment by davmason** [comment]
I have done some research and the behavior of the ResumeRuntime command is by design. We keep track of each port's suspend status (so you can suspend for multiple sessions) and then pause startup until all ports that requested a suspend have marked themselves as resumed.

So if you have a reverse server port and then connect to the default diagnostic port to send ResumeRuntime, it will mark the default diagnostic port as resumed but keep suspended because the reverse server port is still waiting for its ResumeRuntime.

These two functions are the interesting ones, if anyone wants to dig through code: https://github.com/dotnet/runtime/blob/7c00b17be1b2ffb6ed49ad68cf36e9a056323152/src/native/eventpipe/ds-server.c#L259-L285

**Comment by jander-msft** [comment]
> I have done some research and the behavior of the ResumeRuntime command is by design. We keep track of each port's suspend status (so you can suspend for multiple sessions) and then pause startup until all ports that requested a suspend have marked themselves as resumed.
> 
> So if you have a reverse server port and then connect to the default diagnostic port to send ResumeRuntime, it will mark the default diagnostic port as resumed but keep suspended because the reverse server port is still waiting for its ResumeRuntime.
> 
> These two functions are the interesting ones, if anyone wants to dig through code:
> 
> https://github.com/dotnet/runtime/blob/7c00b17be1b2ffb6ed49ad68cf36e9a056323152/src/native/eventpipe/ds-server.c#L259-L285

That wasn't the issue though; the hang problem was occurring from:
1. Accept reverse connection stream
1. Send ApplyStartupHook on default connection
1. Dispose reverse connection stream
1. Accept reverse connection stream (this is where the hang occurred)
1. Send ResumeRuntime on reverse connection
1. Dispose reverse connection stream

[This commit](https://github.com/dotnet/runtime/pull/86813/commits/e6504fe999e6a158a9e3fcba8dabdac0f68b915b) fixed the test that shows changing from using the default connection for the ApplyStartupHook to using the reverse connection stream makes it work.

**Comment by jander-msft** [comment]
It looks like the test legs that are cancelling due to timeouts are because the Helix runs are taking a lot longer than they normally should. I've looked an some of those and all of the work items eventually complete, just longer than the timeout for the AzDO job (which is 2.5 hrs). For example:
- https://dev.azure.com/dnceng-public/public/_build/results?buildId=299126&view=logs&j=b04f8d62-12b9-5ca9-905d-ab402cb4d343&t=714b7eb2-c9c3-572b-df40-4b7c63f4e2cb waited for 2.5 hours and gave up; it's [Helix job](https://[helix.dot.net/api/jobs/189f064b-6f73-446c-ae66-63b785bb03ef/details?api-version=2019-06-17](https://helix.dot.net/api/jobs/189f064b-6f73-446c-ae66-63b785bb03ef/details?api-version=2019-06-17)) took nearly 3 hours to complete.
- https://dev.azure.com/dnceng-public/public/_build/results?buildId=299126&view=logs&j=2a7928c1-3216-5297-550d-59030f19332e&t=db9f5eb1-4373-5f79-4183-560948516f6c waited for 2.5 hours; three of the Helix jobs completed but [one of the jobs](https://[helix.dot.net/api/jobs/e9e0a72e-5c8f-4a85-afc4-c9e087418ee0/details?api-version=2019-06-17](https://helix.dot.net/api/jobs/e9e0a72e-5c8f-4a85-afc4-c9e087418ee0/details?api-version=2019-06-17)) hasn't started processing its work items yet.

The other reported test failures seem to be unrelated to these changes.

Any guidance on what's the path forward from here?

**Comment by davmason** [comment]
> That wasn't the issue though; the hang problem was occurring from:
> 
> 1. Accept reverse connection stream
> 2. Send ApplyStartupHook on default connection
> 3. Dispose reverse connection stream
> 4. Accept reverse connection stream (this is where the hang occurred)
> 5. Send ResumeRuntime on reverse connection
> 6. Dispose reverse connection stream
> 
> [This commit](https://github.com/dotnet/runtime/pull/86813/commits/e6504fe999e6a158a9e3fcba8dabdac0f68b915b) fixed the test that shows changing from using the default connection for the ApplyStartupHook to using the reverse connection stream makes it work.

In step 4 do you accept on the same port as step 1 or a different port? If it's the same port I am thinking to not bother with fixing it. It isn't the best behavior but since we expect to get the ResumeRuntime command on the same connection as the ReverseServer I don't think we need to spend a lot of effort making sure we fail the right way when we receive the command on a different port.

**Comment by jander-msft** [comment]
> > That wasn't the issue though; the hang problem was occurring from:
> > 
> > 1. Accept reverse connection stream
> > 2. Send ApplyStartupHook on default connection
> > 3. Dispose reverse connection stream
> > 4. Accept reverse connection stream (this is where the hang occurred)
> > 5. Send ResumeRuntime on reverse connection
> > 6. Dispose reverse connection stream
> > 
> > [This commit](https://github.com/dotnet/runtime/pull/86813/commits/e6504fe999e6a158a9e3fcba8dabdac0f68b915b) fixed the test that shows changing from using the default connection for the ApplyStartupHook to using the reverse connection stream makes it work.
> 
> In step 4 do you accept on the same port as step 1 or a different port? If it's the same port I am thinking to not bother with fixing it. It isn't the best behavior but since we expect to get the ResumeRuntime command on the same connection as the ReverseServer I don't think we need to spend a lot of effort making sure we fail the right way when we receive the command on a different port.

It's the same port.

**Comment by jander-msft** [comment]
I keep forgetting I can't kick off new builds for this repo, can someone kick one off again for me? Trying to get a fresh PR build merge point and hoping to get through the backed up Helix queues.

**Comment by jander-msft** [comment]
I was hoping to get a new build (not a rerun of the existing build) in order to see if the chrome-DebuggerTests and the build failure on freebsd could be avoided.

**Comment by jander-msft** [comment]
Finally have a build that didn't time out.

I logged a know build issue (https://github.com/dotnet/runtime/issues/87323) because the work items sent to the `(AlmaLinux.8.Amd64.Open)RedHat.7.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:almalinux-8-helix-amd64` Helix queue are failing to be executed due to some Docker issue on the machines.

The [Build Analysis](https://github.com/dotnet/runtime/pull/86813/checks?check_run_id=14141253735) seems to have attributed all of the failures to known issues.

I would like to ask for approvals and this to be merged if there isn't any additional feedback.

**Comment by davmason** [comment]
I'm happy with the current state, @elinor-fung, @lateralusX, or @vitek-karas if you want to take a further look let us know, otherwise I will merge it after end of day Monday.

**Pull request merged**
